packages/shared/src/libs/helpers/check-has-permission/check-has-permission.helper.ts

```diff
@@ -1,17 +1,24 @@
-import { EMPTY_LENGTH } from "../../../libs/constants/empty-length.constant.js";
 import { type PermissionGetAllItemResponseDto } from "../../../modules/permissions/permissions.js";
+import { type Permissions } from "../../types/types.js";
 
 const checkHasPermission = (
-	requiredPermissions: string[],
-	permissions: PermissionGetAllItemResponseDto[],
+	expectedPermissions: Permissions,
+	actualPermissions: PermissionGetAllItemResponseDto[],
 ): boolean => {
-	if (requiredPermissions.length === EMPTY_LENGTH) {
-		return true;
-	}
+	const actuals = new Set(actualPermissions.map((actual) => actual.key));
+	const { alternative, required } = expectedPermissions;
 
-	return requiredPermissions.some((permission) =>
-		permissions.some((userPermission) => userPermission.key === permission),
-	);
+	const hasRequiredPermissions = required
+		? required.every((requiredPermission) => actuals.has(requiredPermission))
+		: true;
+
+	const hasAlternativePermission = alternative
+		? alternative.some((alternativePermission) =>
+				actuals.has(alternativePermission),
+			)
+		: true;
+
+	return hasRequiredPermissions && hasAlternativePermission;
 };
 
 export { checkHasPermission };
```

Comment:
This keeps it centralized and checks the permissions in the same place, this will handle cases where it can accept both X and Y permissions as well  A or B. It keeps the logic for the check inside of the check has permissions.  

You suggested change from the comments is simpler but will fail if we ever get 2 required permissions since it will only check OR, unless you set them separately. as basePermission and additional, which doesn't make it clear.  
I do not think this is a problem for our project since we are currently not using 2 required and an additional anywhere, but for scalability and clarity, the handling for this might be better in first iteration?